qt-models/cylindermodel.cpp

```diff
@@ -486,19 +486,34 @@ static bool show_cylinder(struct dive *dive, int i)
 
 void CylindersModel::updateDive()
 {
-	clear();
-	rows = 0;
+	int i, j;
+	int invisible_rows = 0;
+	int last_row = 0;
 #ifdef DEBUG_CYL
 	dump_cylinders(&displayed_dive, true);
 #endif
-	for (int i = 0; i < MAX_CYLINDERS; i++) {
-		if (show_cylinder(&displayed_dive, i))
-			rows = i + 1;
+	// Find last row
+	for (i = 0; i < MAX_CYLINDERS; i++) {
+		if (!cylinder_none(&displayed_dive.cylinder[i]))
+			last_row = i;
 	}
+	// Move invisible cylinders to the end of the list
+	for (i = j = 0; j <= last_row; i++, j++) {
+		if (!show_cylinder(&displayed_dive, i)) {
+			moveAtPos(i, last_row);
+			invisible_rows++;
+			if (i != last_row) 
+				i--;
+		}
+	}
+	clear();
+	rows = last_row + 1 - invisible_rows;
 	if (rows > 0) {
 		beginInsertRows(QModelIndex(), 0, rows - 1);
 		endInsertRows();
 	}
+	if (current_dive)
+		copy_dive(&displayed_dive, current_dive);
```

Comment:
This has a few interesting side effects that I'm not sure you want.
The idea is that the displayed_dive is what you see, the current_dive is what's in our data storage.
So at any time a partial edit could be active that hasn't been accepted, yet. So no, please don't do that.